include/exiv2/slice.hpp

```diff
@@ -33,689 +33,548 @@
-             */
-            value_type& unsafeAt(size_t index) noexcept
-            {
-                return data_[index];
-            }
+        /// Return an iterator to the first element beyond the slice.
+        ///
+        /// Note: this function is not available when `container` is a constant
+        /// type.
+        ///
+        /// @throw  This function does not throw anything, as long as
+        ///     `container::begin()` and `iterator++` don't throw.
+#ifdef PARSED_BY_DOXYGEN
+        iterator
+#else
+        template <typename Dummy = container>
+        typename std::enable_if<!std::is_const<Dummy>::value, iterator>::type
+#endif
+        end()
+        {
+            RETURN_ITERATOR(begin, this->end_);
+        }
 
-            const value_type& unsafeAt(size_t index) const noexcept
-            {
-                return data_[index];
-            }
+        /// Return a constant iterator to the first element beyond the slice.
+        ///
+        /// @throw  This function does not throw anything, as long as
+        ///     `container::begin()` and `iterator++` don't throw.
+        const_iterator end() const
+        {
+            RETURN_ITERATOR(begin, this->end_);
+        }
 
-            /*!
-             * Obtain an iterator (=pointer) at the position of the element with
-             * the given index in the container.
-             *
-             * @throw nothing
-             */
-            iterator unsafeGetIteratorAt(size_t index) noexcept
-            {
-                return data_ + index;
-            }
+#ifndef PARSED_BY_DOXYGEN
+#undef RETURN_ITERATOR
+#endif
+
+        /// Construct a sub-slice of this slice with the given bounds. The bounds
+        /// are evaluated with respect to the current slice.
+        ///
+        /// Note: the non-const overload is only available when the container is
+        /// also not constant.
+        ///
+        /// Example:
+        /// ~~~~~~~~~~~~~~~~~~~~~~~~{.cpp}
+        /// std::vector<int> vec = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+        /// slice<std::vector<int>> center = Slice(vec, 3, 7);
+        /// // center_vals = {3, 4, 5, 6}
+        /// auto middle = center_vals.subSlice(1, 3);
+        /// // middle = {4, 5}
+        /// // the following throws:
+        /// center_vals.subSlice(1, 5);
+        /// ~~~~~~~~~~~~~~~~~~~~~~~~
+        ///
+        /// @param[in] begin  First element in the new slice.
+        /// @param[in] end  First element beyond the new slice.
+        ///
+        /// @throw  std::out_of_range when begin or end are out of bounds of the
+        ///     current slice
+#ifdef PARSED_BY_DOXYGEN
+        Slice<container>
+#else
+        template <typename Dummy = container>
+        typename std::enable_if<!std::is_const<Dummy>::value, Slice<container> >::type
+#endif
+        subSlice(size_t begin, size_t end)
+        {
+            this->calculateSubSliceBounds(begin, end);
+            return Slice(cont_, begin, end);
```

Comment:
@piponazo stupid question: why is this working? This should only compile with C++17.

Reply:
No idea ... I did not implement this! 😝 

What is the "since-C++17" feature you are using here?

Reply:
Afaik constructors cannot deduce template parameters before C++17, that's why I added the `makeSlice()` functions. So this shouldn't compile... But it does and codecov even says that it is executed. Very odd.

Reply:
According to this [link](https://gcc.gnu.org/projects/cxx-status.html) this feature is implemented in GCC > 7 and in CI we are using 5.4:

https://travis-ci.org/Exiv2/exiv2/jobs/500930437

The only thing I can say is that I am pretty sure we are compiling with `-std=c++11`

Reply:
Back to the drawing board again. I am starting to regret adding this code to exiv2, it's already causing problems and it hasn't even been really used :(

Reply:
I've asked around and apparently this is a C++ "feature": if you construct a template class from inside the same template class, then the current template parameters will be implicitly inserted for you. So it looks like template type argument deduction for constructors, but it's actually something completely different.